README
contributor: Wang Ruizhe 521370910147

#basic information
This is a cross road traffic game for users. 
In this game, users decide on themselves several features concerining the scenario. 
The way to win is to avoid traffic accidents and let all the cars pass the crossing successfully.
If cars crash into each other, the user loses without seeing the scene of the accident.
Also, users can decide the probability that a car does not stop at the traffic light, which will add risks to the game.
At the end of the game, the plates of the cars that disobey traffic rules will be printed out.

#operations
The following things are supposed to be inputted by the users themselves:

1.the width of the road
Since the width of the cars are pre-defined by the designer, the width of the road is supposed to be larger than 0.4,which is double the width of the car. 
The designer recommend 1 as a good choice.(described in pixels)

2.the time lapses of each colored light(red,orange,green),
This decision will greatly affect whether or not the user will win.
In the designer's opinion, the time lapses should be identical to those in real life, not too long or too short.
After trials and errors, the designer thinks 60 for the red light, 60 for the green light, 1 for the orange light can be a good choice.(the numbers represent time units,not the actual seconds)

3.the total number of cars
Users decide how many cars in total will appear on the roads. the total number will be randomly distributed to each lane.
Since it's happier to have the road less crowded, 10 to 15 cars can be good, but changes can be made if the time lapses are different.

4.the probability that a car does not stop at the red light
As is mentioned above, this is an option that will add risks to the game.
The larger it is, the more likely the user will lose.

#required algorithms
1.the motion of cars
Take one lane as an example
if: the number of the cars isn't zero, place the cars on their initial positions.
Also, give each of them a "possibility" to be compared later.
if: the "possibility" is smaller than p(the inputted possibility), move forward by one unit.
else:if the car is not the first at the crossroad,
           if the first car hasn't stopped or the car has moved beyond the cross, move forward by one unit.
           else, remain at its original position.
       else(if the car is the first at the crossroad),
           if the lightcolor is green, move forward by one unit
           else,remain at its original position.
after each status, the cars are deleted and drawn again.

cars on all lanes repeat this algorithm to move.

p.s. about crashing
after each time a set of cars are completely drawn
for car1 and car2
if the X of car1 at the left bottom is smaller than the X of car2 at the right bottom and the X of car1 at the right bottom is larger than the X of car2 at the left bottom
     if the Y of car1 at the left bottom is smaller than the Y of car2 at the top left corner and the Y of car1 at the top left corner is larger than the Y of car2 at the left bottom
     the cars crash, the game ends.

This process repeats 4 times, for each kind of combination.

2.the change of color of the traffic lights
lightcolor1 stands for the conditions of two lights; lightcolor2 stands for the other two.
input: the time lapses for each color
create a vector with r zeros, g ones and o twos in order, add it to itself according to the times it's needed.
The color of lightcolor1 is decided by vector(mi) and the color of lightcolor2 is decided by vector(mj).
mi starts from 1, mj starts from r+1, each stands for the element at the corresponding place in the vector.
If the number in the vector is 0,the color is red; if the number is 1, green; if the number is 2, orange.
Add 1 to mi and mj each time a set of cars are completely drawn so as to fetch the next number in the vector, the colors are then decided again and again according to the numbers.
after each status, the lights are deleted and drawn again.


3.registration plates
X is an empty vector
function X=plate(X,l)
   if l=1 then give it a letter from A to Z; store it in X
   else l=l-1;X=plate(X,l)(recall the function); provide a letter from A to Z or a number from 0 to 9; store it in X
Repeat the function as many times as the user wants(n times for example).
The permutations are stored in X.

#unique parts
1.The time unit is 0.3 seconds and the moving unit every time unit is 0.1.
2.The cars are controlled by the light across the middle cross, say, the one farther away.
3.At the end of the game, if the user loses, all the cars disappear and he will receive a "Game over!" in the command window.
If he's successful, the cars will also disappear when they all go through the cross, and no sentence will be printed out.
Whether or not he's successful, the plates of the "bad cars" that run the red lights will be printed out in the command window.
In my understanding, running orange lights are also bad behaviors, so these cars are treated the same as those who run red lights.
4.If two car accidents happen at the same time, the user will receive 2 "Game over!"s to make him clear about the heavy loss.

#files
1.p1 
the main script, add up all the functions and make certain changes

2.crossroad 
the function that draws the crossroad

3.drawing3
the function that draws the lights and cars

4.numberofcars
the function that randomly generatesthe number of cars for each lane

5.plate
the function that generates the plates

6.showthebadcars
the function that generates the plates of the "bad cars" that disobeys traffix rules

7.Changelog
record my progress

8.README

Changelog
#Changelog

##[0.0.1]-2021-10-08

###Added:
the drawing of cars
generation of each digit of the plate

###Changed:
put the drawings of lights in one function

##[0.0.2]-2021-10-15

###Added：
the generation of all the plates needed
decide whether a car runs the red light(or the orange light) and display its plate
randomly distribute the number of cars to each lane
animate the movement of cars
the display of "Game over!" when a user loses the game
the "probabiity" for each car

###Changed:
put the drawings of lights and cars in one function
change the moving unit from 1 to 0.1 and the pausing time from 1 to 0.3 to make the scenerio smoother
make the lights shine by letting it read numbers from a vector, instead of using an infinite loop, which makes it unable to enter the next step
the cars and lights are drawn again every time unit, instead of adding a car to the front each time
link the "probability" and the p to decide the motion of cars
the way of generating random car numbers

###Removed
the infinite loop for lights
the loop that adds a new car to the front every time

drawing3.m
function [lightcolor1,lightcolor2]=drawing3(w,O,P,I,L,lightcolorall,mi,mj,n1,n2,n3,n4)
n=1;r=1;s=1;h=1;
if lightcolorall(mi)==0
    lightcolor1='r';
end
if lightcolorall(mi)==1
    lightcolor1='g';
end
if lightcolorall(mi)==2
    lightcolor1='#EDB120';
end
if lightcolorall(mj)==0
    lightcolor2='r';
end
if lightcolorall(mj)==1
    lightcolor2='g';
end
if lightcolorall(mj)==2
    lightcolor2='#EDB120';
end
R=[];W=[];Q=[];E=[];
if n1~=0
    while n<=n1
        car1=rectangle('Position',[3+3*w/4-0.1,O(n),0.2,0.3],'FaceColor',[0.9290 0.6940 0.1250],'Edgecolor','k','Linewidth',0.5);
        n=n+1;
        R=[R,car1];
    end
end
if n2~=0
    while r<=n2
        car2=rectangle('Position',[P(r),3+w/4-0.1,0.3,0.2],'FaceColor',[0.3010 0.7450 0.9330],'Edgecolor','k','Linewidth',0.5);
        r=r+1;
        W=[W,car2];
    end
end
if n3~=0
    while s<=n3
        car3=rectangle('Position',[3+w/4-0.1,6+w-I(s),0.2,0.3],'FaceColor',[0.4940 0.1840 0.5560],'Edgecolor','k','Linewidth',0.5);
        s=s+1;
        Q=[Q,car3];
    end
end
if n4~=0
    while h<=n4
        car4=rectangle('Position',[6+w-L(h),3+3*w/4-0.1,0.3,0.2],'FaceColor',[0.8500 0.3250 0.0980],'Edgecolor','k','Linewidth',0.5);
        h=h+1;
        E=[E,car4];
    end
end
rectangle('Position',[2.8 2.8+w 0.4 0.4],'Curvature',[1 1],'FaceColor',lightcolor1,"EdgeColor",lightcolor1);
rectangle('Position',[2.8+w 2.8+w 0.4 0.4],'Curvature',[1 1],'FaceColor',lightcolor2,"EdgeColor",lightcolor2);
rectangle('Position',[2.8 2.8 0.4 0.4],'Curvature',[1 1],'FaceColor',lightcolor2,"EdgeColor",lightcolor2);
rectangle('Position',[2.8+w 2.8 0.4 0.4],'Curvature',[1 1],'FaceColor',lightcolor1,"EdgeColor",lightcolor1);
pause(0.3);
delete(R);delete(W);delete(Q);delete(E);delete(rectangle);
end


p1.m
%input according to the hints
w=input('the width of the road: ');
g=input('the time lapse of the green light: ');
r=input('the time lapse of the red light: ');
o=input('the time lapse of the orange light: ');
n=input('the total number of cars: ');
p=input('the probability that a car does not stop at the red light: ');
[n1,n2,n3,n4]=numberofcars(n);%randomly decide the number of cars on each route
U=[n1,n2,n3,n4];
crossroad(w);%draw the crossroad
j=6;
X=[];
%generate plates using recursion
for i=1:n
    X=plate(X,j);
end
PLATE=X;
lightcolorall=[];
%create a vector according to different time lapses
for d=1:r
    lightcolorall=[lightcolorall,0];
end
for d=1:g
    lightcolorall=[lightcolorall,1];
end
for d=1:o
    lightcolorall=[lightcolorall,2];
end
usefulnumber=1;
neededtimes=lightcolorall;
%create a vector large enough so as the lights will shine till the end
while usefulnumber<=(33+10*w+5*max(U))/(r+g+o)+10%probably larger than needed, just to be safe
    lightcolorall=[lightcolorall,neededtimes];
    usefulnumber=usefulnumber+1;
end
%vectors to store values concerning locations, possibilities and plates
O=[];P=[];I=[];L=[];prob1=[];prob2=[];prob3=[];prob4=[];PLATE1=[];PLATE2=[];PLATE3=[];PLATE4=[];
conditionall=[];
%place the cars' original states accordingly on each lane
%Also,give each car a plate and its "probability" to run a light(compared with "p" later)
if n1~=0
    for d=1:n1
        O=[O,(d-1)*(-0.5)];
        plate1=PLATE(6*d-5:6*d);
        PLATE1=[PLATE1,plate1];
        prob1=[prob1,rand];
    end
    conditionall=[conditionall,O(n1)];%store the original values for later use
end
if n2~=0
    for d=1:n2
        P=[P,(d-1)*(-0.5)];
        plate2=PLATE(6*n1+6*d-5:6*n1+6*d);
        PLATE2=[PLATE2,plate2];
        prob2=[prob2,rand];
    end
    conditionall=[conditionall,P(n2)];
end
if n3~=0
    for d=1:n3
        I=[I,(d-1)*(-0.5)];
        plate3=PLATE(6*n1+6*n2+6*d-5:6*n1+6*n2+6*d);
        PLATE3=[PLATE3,plate3];
        prob3=[prob3,rand];
    end
    conditionall=[conditionall,0.3+I(n3)];
end
if n4~=0
    for d=1:n4
        L=[L,(d-1)*(-0.5)];
        plate4=PLATE(6*n1+6*n2+6*n3+6*d-5:6*n1+6*n2+6*n3+6*d);
        PLATE4=[PLATE4,plate4];
        prob4=[prob4,rand];
    end
    conditionall=[conditionall,L(n4)-0.3];
end
mi=1;mj=r+1;
j=1;r=1;s=1;h=1;
flagall=false;
%vectors to store the plates of "bad cars" who disobey traffic rules
badcars1=[];badcars2=[];badcars3=[];badcars4=[];
axis([0 6+w 0 6+w]);%define the boundaries
axis off%erase the axis
while min(conditionall)<=3+w
    conditionall=[];
    if flagall==true
        break
    end
    [lightcolor1,lightcolor2]=drawing3(w,O,P,I,L,lightcolorall,mi,mj,n1,n2,n3,n4);%call the function that draws lights and cars
    flag1=0;flag2=0;flag3=0;flag4=0;
    if n1~=0%if the number distributed is zero, don't do the things considering the lane
        for j=1:n1
            if prob1(j)<=p&&O(j)>=2.695&&O(j)<=2.705%decide whether the car is the first at the crossing
                if lightcolor2=='g'%if the light is green, go ahead as it pleases
                    O(j)=O(j)+0.1;
                else
                    O(j)=O(j)+0.1;%if not a green light and it goes ahead, store it as a bad car
                    nnn=j;
                    badcars1=[badcars1,nnn];
                end
            else
                if O(j)<=2.695 || O(j)>=2.705
                    if flag1~=1 || O(j)>=2.705%if it's behind the cross,decide whether the car in the front has stopped;if not, go ahead
                        O(j)=O(j)+0.1;
                    else
                        O(j)=O(j);
                    end
                else
                    if lightcolor2=='g'%decide the motion of the first car
                        O(j)=O(j)+0.1;
                    else
                        O(j)=O(j);
                        flag1=1;%control the cars behind to stop
                    end
                end
            end
        end
        conditionall=[conditionall,O(j)];%store the final position
    end
    if n2~=0
        for r=1:n2
            if prob2(r)<=p&&P(r)>=2.695&&P(r)<=2.705
                if lightcolor1=='g'
                    P(r)=P(r)+0.1;
                else
                    P(r)=P(r)+0.1;
                    rrr=r;
                    badcars2=[badcars2,rrr];
                end
            else
                if P(r)<=2.695 || P(r)>=2.705
                    if flag2~=1 || P(r)>=2.705
                        P(r)=P(r)+0.1;
                    else
                        P(r)=P(r);
                    end
                else
                    if lightcolor1=='g'
                        P(r)=P(r)+0.1;
                    else
                        P(r)=P(r);
                        flag2=1;
                    end
                end
            end
        end
        conditionall=[conditionall,P(r)];
    end
    if n3~=0
        for s=1:n3
            if prob3(s)<=p&&I(s)>=2.995&&I(s)<=3.005
                if lightcolor2=='g'
                    I(s)=I(s)+0.1;
                else
                    I(s)=I(s)+0.1;
                    sss=s;
                    badcars3=[badcars3,sss];
                end
            else
                if I(s)<=2.995 || I(s)>=3.005
                    if flag3~=1 || I(s)>=3.005
                        I(s)=I(s)+0.1;
                    else
                        I(s)=I(s);
                    end
                else
                    if lightcolor2=='g'
                        I(s)=I(s)+0.1;
                    else
                        I(s)=I(s);
                        flag3=1;
                    end
                end
            end
        end
        conditionall=[conditionall,0.3+I(s)];
    end
    if n4~=0
        for h=1:n4
            if prob4(h)<=p&&L(h)>=2.995&&L(h)<=3.005
                if lightcolor1=='g'
                    L(h)=L(h)+0.1;
                else
                    L(h)=L(h)+0.1;
                    hhh=h;
                    badcars4=[badcars4,hhh];
                end
            else
                if L(h)<=2.995 ||L(h)>=3.005
                    if flag4~=1 || L(h)>=3.005
                        L(h)= L(h)+0.1;
                    else
                        L(h)= L(h);
                    end
                else
                    if lightcolor1=='g'
                        L(h)= L(h)+0.1;
                    else
                        L(h)= L(h);
                        flag4=1;
                    end
                end
            end
        end
        conditionall=[conditionall,L(h)-0.3];
    end
    %use the positions of bottoms to decide whether crashes happen
    if n2~=0&&n1~=0
        for r=1:n2
            for j=1:n1
                if P(r)+0.3>=3+3*w/4-0.1&&P(r)<=3+3*w/4-0.1+0.2
                    if 3+w/4-0.1<=O(j)+0.3&&3+w/4-0.1+0.2>=O(j)
                        disp("Game Over!");flagall=true;
                    end
                end
            end
        end
    end
    if n2~=0&&n3~=0
        for r=1:n2
            for s=1:n3
                if P(r)+0.3>=3+w/4-0.1&&P(r)<=3+w/4-0.1+0.2
                    if 3+w/4-0.1<=6+w-I(s)+0.3&&3+w/4-0.1+0.2>=6+w-I(s)
                        disp("Game Over!");flagall=true;
                    end
                end
            end
        end
    end
    if n4~=0&&n3~=0
        for h=1:n4
            for s=1:n3
                if 6+w-L(h)+0.3>=3+w/4-0.1&&6+w-L(h)<=3+w/4-0.1+0.2
                    if 3+3*w/4-0.1<=6+w-I(s)+0.3&&3+3*w/4-0.1+0.2>=6+w-I(s)
                        disp("Game Over!");flagall=true;
                    end
                end
            end
        end
    end
    if n4~=0&&n1~=0
        for h=1:n4
            for j=1:n1
                if  6+w-L(h)+0.3>=3+3*w/4-0.1&&6+w-L(h)<=3+3*w/4-0.1+0.2
                    if 3+3*w/4-0.1<=O(j)+0.3&&3+3*w/4-0.1+0.2>=O(j)
                        disp("Game Over!");flagall=true;
                    end
                end
            end
        end
    end
    mi=mi+1;mj=mj+1;%switch the lights to the next status
end
showthebadcars(PLATE,badcars1,badcars2,badcars3,badcars4,n1,n2,n3);%call a function to display the plates of the "bad cars"



